core-bundle/src/EventListener/PreviewToolbarListener.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\CoreBundle\Routing\ScopeMatcher;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\Routing\RouterInterface;
+use Twig\Environment as TwigEnvironment;
+use Twig\Error\Error as TwigError;
+
+/**
+ * Injects the back end preview toolbar on any response within the /preview.php
+ * entry point.
+ *
+ * The onKernelResponse method must be connected to the kernel.response event.
+ *
+ * The toolbar is only injected on well-formed HTML with a proper </body> tag,
+ * so is never included in sub-requests or ESI requests.
+ */
+class PreviewToolbarListener
+{
+    /**
+     * @var ScopeMatcher
+     */
+    private $scopeMatcher;
+
+    /**
+     * @var string
+     */
+    private $previewScript;
+
+    /**
+     * @var TwigEnvironment
+     */
+    private $twig;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    public function __construct(string $previewScript, ScopeMatcher $scopeMatcher, TwigEnvironment $twig, RouterInterface $router)
+    {
+        $this->previewScript = $previewScript;
+        $this->scopeMatcher = $scopeMatcher;
+        $this->twig = $twig;
+        $this->router = $router;
+    }
+
+    public function onKernelResponse(ResponseEvent $event): void
```

Comment:
The method should be changed to `__invoke()` so we can simplify the event registration (see #1125).